pallets/collator-selection/src/lib.rs

```diff
@@ -635,8 +645,18 @@ pub mod pallet {
 
         /// Reset the performance map to the currently active validators at 0 blocks
         pub fn reset_collator_performance() {
-            <BlocksPerCollatorThisSession<T>>::remove_all(None);
             let validators = T::ValidatorRegistration::validators();
+            let validators_len = validators.len() as u32;
+            let mut clear_res = <BlocksPerCollatorThisSession<T>>::clear(validators_len, None);
```

Comment:
This is a bit of an interesting case and at odds with the intention of this new API

1. This fn is only called once on a session boundary. If there is storage left over, no attempts are made in the next block to remove the remaining ones ( this fn would need to be called again, needs a rework on how it gets called and potentially a new marker storage item to keep track of it ).
If this ever gets interrupted midway because the block is full - i.e. the `break` code branch - , it'll lead to incorrect block performance accounting in the next session, the loop logic won't prevent this.

2. It does not matter because a session change block always has the full block to work with  and we don't come close to exhausting a block with the session change logic for the current amount of validators. 
I have no concrete weight number for the whole process tho ( especially with nimbus enabled ), so this is just an educated guess atm. We might need to revisit the weight totals on session change before enabling parachain staking on KMA.

Your implementation looks correct tho.

Reply:
Right, I forgot that this will get only at session boundaries, but my thinking was the same for having an extra storage item where we store the cursor.

Reply:
Actually we can keep deleting in the `on_initialize` hook.
pallet_session's on_init hook gets called first and that can set the cursor, then collator_selection's on_init can continue in the same block and in the next.